server/src/main/java/org/elasticsearch/tasks/TaskCancellationService.java

```diff
@@ -425,7 +425,7 @@ public void cancelChildRemote(TaskId parentTask, long childRequestId, Transport.
                 reason
             );
             final CancelChildRequest request = CancelChildRequest.createCancelChildRequest(parentTask, childRequestId, reason);
-            transportService.sendRequest(childNode, CANCEL_CHILD_ACTION_NAME, request, TransportRequestOptions.EMPTY, NOOP_HANDLER);
+            transportService.sendRequest(childConnection, CANCEL_CHILD_ACTION_NAME, request, TransportRequestOptions.EMPTY, NOOP_HANDLER);
```

Comment:
This fixes the problem.
The `transportService` (its inner connection manager) here doesn't know about the `childConnection` to the `childNode`, because the `childNode` is a remote node, and hence the`childConnection` is managed differently: https://github.com/elastic/elasticsearch/blob/2822fa407218fdf566bcc33424d44c9f44c2d73a/server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java#L30

Reply:
Got a bit confused reading this, but I understand that the `childNode` as a DiscoveryNode is not known to the local `transportService` if it's a remote node. So it doesn't send the cancellation. You'd like to use immediately the `childConnection`.

Indeed when we first did https://github.com/elastic/elasticsearch/pull/92588 , I do not remember specifically tackling remote connections. So this fix makes sense.

However, I see the CI failed on the test you introduced. Can you fix it? Just to make sure we're not missing anything.

Reply:
It was [explicitly changed](https://github.com/elastic/elasticsearch/pull/92588/commits/c9f685d5bed8de8a6d6aba3b98969327cddffe01#diff-6bb936c7c30ca4f5caf33cb3bdd013d0cbf001ce5032a0be461ee91e0fc26af3R409) to use `childNode` based on a [review comment](https://github.com/elastic/elasticsearch/pull/92588/files#r1088763415).

I am not entirely sure the comment is valid, we should normally leave the connection.

Reply:
>However, I see the CI failed on the test you introduced. Can you fix it? Just to make sure we're not missing anything.

@kingherc 
I've narrowed down the issue in the test failure to:
```
diff --git a/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java b/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
index 93c9f63fc5e..0a9c3d9dd87 100644
--- a/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
@@ -80,6 +80,7 @@ import java.util.function.Function;
 import java.util.function.Supplier;

 import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.spy;

 /**
  * A mock delegate service that allows to simulate different network topology failures.
@@ -314,7 +315,7 @@ public class MockTransportService extends TransportService {
         if (MockTaskManager.USE_MOCK_TASK_MANAGER_SETTING.get(settings)) {
             return new MockTaskManager(settings, threadPool, taskHeaders);
         } else {
-            return new TaskManager(settings, threadPool, taskHeaders, tracer);
+            return spy(new TaskManager(settings, threadPool, taskHeaders, tracer));
         }
     }
```
onto the `main` branch.
The test failure reproduces with:
```
./gradlew ':modules:repository-azure:internalClusterTest' --tests "org.elasticsearch.repositories.azure.AzureBlobStoreRepositoryTests.testRequestStats" -Dtests.seed=58054329A825A1B9 -Dtests.locale=en-AU -Dtests.timezone=Europe/Prague -Druntime.java=22
```
It always generates an `OutOfMemoryError`. I have no idea what the problem might be, but it's not related to the code change in the main code.

In https://github.com/elastic/elasticsearch/pull/109440/commits/4f78f71149f7527cd5ab6654f37bcb92e0b70571 I've made the test code spy on the mock task manager only, which is not used in the `AzureBlobStoreRepositoryTests` and that should pass all CI.

@kingherc Can you take another look, please?

Reply:
Ah thanks @henningandersen for reminding us of the comment:

> I would prefer to use the node instead of the connection here, that would avoid issues if the connection is dropped for some reason.

I guess if I had to choose over missing to send (rarely) some cancellations, vs making sure we send cancellations to remote clusters, I'd prefer the latter :+1: 

I think I'm a bit skeptical now about using the `spy`. spy() introduces a memory overhead as mockito then remembers all invocations, and that's why we probably saw the OOM in CI. And maybe we should try to avoid it if we can. I think if we correct a thing in the test (see my other comment), we can remove the `spy` altogether.